pkg/hubble/observer/local_observer.go

```diff
@@ -300,6 +303,8 @@ nextEvent:
 					continue nextEvent
 				}
 			}
+			// Remove fields
+			mask.Filter(ev)
```

Comment:
Ideas:
1. Invalidate flows from ring buffer - you won't be able to request the ring buffer contents multiple times
2. Make a (partial) deep copy of a flow before clearing parts of its contents
3. [preferred] Copy only relevant fields before sending (no memory allocation overhead, can get rid of new import)

What do you think?

Reply:
AFAIK 1. (Invalidate flows from ring buffer) would not be practical because then a given flow can be observed by only one client. Option 3 seems ideal indeed.

Reply:
My opinion is the following:

> 1. Invalidate flows from ring buffer - you won't be able to request the ring buffer contents multiple times

For me this is a non-starter. We can't assume anything about how many times a flow is observed

> 2. Make a (partial) deep copy of a flow before clearing parts of its contents

This probably has a simpler implementation, but would incur a higher performance cost compared to 3

> 3. [preferred] Copy only relevant fields before sending (no memory allocation overhead, can get rid of new import)

This is indeed the preferred approach. It will still have some overhead to create a new empty object and then copy over the relevant fields.

As far as actual implementation, it's probably a good idea to have a `sync.Pool` so we limit the impact of new allocation the best we can.

Reply:
I agree, option 3 is best, but I would suggest doing option 2 and benchmarking it. It may be relatively insignificant cost to copy the struct and then remove fields since many of the fields are pointers and slices anyways, meaning we're really just copying a few extra pointers and removing them afterwards. It may even be faster since it's probably just a memcpy of the struct and unassigning a few fields, whereas creating a new struct and filling fields is potentially more total operations.

Reply:
Of course, data driven approach is great here.

Reply:
Unfortunately I was unable to get copying from one `proto` field  to another without allocation. Other solutions I've seen used mapping of `proto->go` field names. The main problem of `[not_so_]shallow_copy` is copying strings [here](https://github.com/protocolbuffers/protobuf-go/blob/master/internal/impl/convert.go#L324).

Bench results:
```
name \ time/op   no_mask      no_copy      deep_copy    shallow_copy
RelayGetFlows-8  19.4µs ±14%  12.3µs ±15%  16.1µs ±16%  13.9µs ±21%

name \ alloc/op  no_mask      no_copy      deep_copy    shallow_copy
RelayGetFlows-8  11.8kB ± 0%  6.0kB ± 0%   7.2kB ± 0%   6.0kB ± 0%

name \ allocs/op no_mask      no_copy      deep_copy    shallow_copy
RelayGetFlows-8  182 ± 0%     113 ± 0%     124 ± 0%     121 ± 0%
```

Reply:
Can you provide memory benchmark results too now that we're copying objects when masking is enabled?

Reply:
What do you mean? Do you need anything more than `alloc/op` and `allocs/op` from my previous comment?

Reply:
Ah yes, I didn't quite notice that, I thought it was just a CPU benchmark, I didn't notice each row had a different header.

Reply:
@AwesomePatrol is the underlying code for this benchmark available anywhere?

Reply:
Sorry for the confusion, but this is just `BenchmarkRelayGetFlowsWithFieldMask` and `BenchmarkRelayGetFlowsWithoutFieldMask` run on different commits in this chain:
`no_mask` is `BenchmarkRelayGetFlowsWithoutFieldMask` on `HEAD` (`cff0a8f0cb3db44387569154e3ed342d3f8152f0`)
`no_copy` is `BenchmarkRelayGetFlowsWithFieldMask` on `HEAD~4` (`8aee1915bb938aeb9d236a2451e4135a3317322b`)
`deep_copy` is `BenchmarkRelayGetFlowsWithFieldMask` on `HEAD~3` (`270c456debd120f2571b6978f3cc6f502a2823cf`)
`shallow_copy` is `BenchmarkRelayGetFlowsWithFieldMask` on `HEAD` (`cff0a8f0cb3db44387569154e3ed342d3f8152f0`)

Everything is published to this pull request